### PR TITLE
Add test to check incorrect output for kimchi backend

### PR DESCRIPTION
**Pull request by katat**
https://github.com/zksecurity/noname/issues/58

The kimchi prover/ verifier already cover the public output constraints. Here we add a test to ensure this is the case.

### CODE REVIEW

File: src/backends/kimchi/prover.rs [hunk 1]
```diff
@@ -233,3 +233,59 @@ impl VerifierIndex {
         .wrap_err("kimchi: failed to verify the proof")
     }
 }
+
+mod testing {
```
Comment (mimoo) [inline src/backends/kimchi/prover.rs:237]: `#[cfg(test)]`

File: src/backends/kimchi/prover.rs [hunk 2]
```diff
@@ -233,3 +233,59 @@ impl VerifierIndex {
...
+        let witness = generated_witness.all_witness.to_kimchi_witness();
+
+        let (prover_index, _) = backend.compile_to_indexes().unwrap();
+
+        // correct output value
+        let mock_public_entries = [input_val, res_val];
+        prover_index.verify(&witness, &mock_public_entries).unwrap();
+
+        // incorrect output value
+        let mock_public_entries = [input_val, VestaField::from(1)];
```
Comment (mimoo) [inline src/backends/kimchi/prover.rs:283]: maybe `res_val + 1` would ensure that it fails :D

File: src/backends/kimchi/prover.rs [hunk 3]
```diff
@@ -233,3 +233,59 @@ impl VerifierIndex {
...
+        let input_var = backend.add_public_input(Value::Constant(input_val), span);
+
+        let output_var = backend.add_public_output(Value::PublicOutput(None), span);
+
+        // 1 + 1 = 2
+        let const_val = VestaField::from(1);
+        let res_val = input_val + const_val;
+        let res_var = backend.add_const(&input_var, &const_val, span);
+        let public_outputs = Some(Var::new_cvar(ConstOrCell::Cell(output_var), span));
+        let returned_cells = Some(Vec::from([res_var]));
```
Comment (mimoo) [inline src/backends/kimchi/prover.rs:265]: I think it would have been nicer to compile something like this instead to make sure the end-to-end flow doesn't have a bug (and then it's easier to check any backend with it as well):

```rust
fn main(pub public_input: Field, private_input: Field) -> Field {
    let xx = private_input + public_input;
    assert_eq(xx, 2);
    let yy = xx + 6;
    return yy;
}
```

the second thing is that... the verify function doesn't check the permutation, so it's not going to detect the bug (if there is one) :D I believe we actually have to create a proof

Comment (mimoo) [reply src/backends/kimchi/prover.rs:265]: so something like that, roughly:

```rust
let code = r#"fn main(pub public_input: Field, private_input: Field) -> Field {
    let xx = private_input + public_input;
    assert_eq(xx, 2);
    let yy = xx + 6;
    return yy;
}";

let mut sources = Sources::new();
let mut tast = TypeChecker::new();
let this_module = None;
let _node_id = typecheck_next_file(
    &mut tast,
    this_module,
    &mut sources,
    "public_output.no".to_string(),
    code.clone(),
    0,
)
.unwrap();

let kimchi_vesta = KimchiVesta::new(false);
let compiled_circuit = compile(&sources, tast, kimchi_vesta)?;

let (prover_index, verifier_index) = compiled_circuit.compile_to_indexes().unwrap();

let generated_witness = generate_witness(
    &self.compiled_circuit,
    sources,
    public_inputs,
    private_inputs,
)?;

// TODO: modify the public output part of the witness

let witness = generated_witness.all_witness.to_kimchi_witness();

// verify the witness
self.index
    .verify(&witness, &generated_witness.full_public_inputs)
    .unwrap();

// create proof
let proof =
    ProverProof::create::<BaseSponge, ScalarSponge>(&GROUP_MAP, witness, &[], &self.index)
        .into_diagnostic()
        .wrap_err("kimchi: could not create a proof with the given inputs")?;

kimchi::verifier::verify::<Curve, BaseSponge, ScalarSponge, OpeningProof<Curve>>(
    &GROUP_MAP,
    &self.index,
    &proof,
    &full_public_inputs,
)
.into_diagnostic()
.wrap_err("kimchi: failed to verify the proof")    
```

**Comment by katat** [comment]
I had a check, the verify function from the ProverIndex seems also including the permutation check? 

```rust
for col in 0..PERMUTS {
  let wire = gate.wires[col];

  if wire.col >= PERMUTS {
      return Err(GateError::Custom {
          row,
          err: format!("a wire can only be connected to the first {PERMUTS} columns"),
      });
  }

  if witness[col][row] != witness[wire.col][wire.row] {
      return Err(GateError::DisconnectedWires(
          Wire { col, row },
          Wire {
              col: wire.col,
              row: wire.row,
          },
      ));
  }
}
```

Once the `assert_eq` constraint for the output is added, this check ^ seems already covering without the need to generate proof. (The refactored test demonstrates this fact)

File: examples/public_output.asm [hunk 4]
```diff
@@ -5,5 +5,8 @@ DoubleGeneric<1>
 DoubleGeneric<1,1,-1>
 DoubleGeneric<1,0,0,0,-2>
 DoubleGeneric<1,0,-1,0,6>
+DoubleGeneric<1,-1>
 (0,0) -> (2,1)
+(1,0) -> (5,0)
```
Comment (mimoo) [inline examples/public_output.asm:10]: shouldn't it be `(0,0) -> (5,0)`?

Comment (mimoo) [reply examples/public_output.asm:10]: oh no, that's the public input, `(1, 0)` is the public output :D

File: src/backends/kimchi/prover.rs [hunk 5]
```diff
@@ -233,3 +233,113 @@ impl VerifierIndex {
...
+        // this is the gate contains the output var
+        assert_eq!(witness[0][1], VestaField::from(output_val));
+        // this is the assert_eq gate which result var and the output var
+        assert_eq!(witness[0][gate_count - 1], VestaField::from(output_val));
+
+        // should pass the sanity check
+        prover_index
+            .index
+            .verify(&witness, &full_public_inputs)
+            .unwrap();
```
Comment (mimoo) [inline src/backends/kimchi/prover.rs:323]: can you also check that this returns an error as well if we modify the pub output then? (if what you say is correct about the permutation I guess this should work as well)

File: src/backends/kimchi/prover.rs [hunk 6]
```diff
@@ -233,3 +233,113 @@ impl VerifierIndex {
...
+
+        assert_eq!(
+            generated_witness.full_public_inputs,
+            vec![
+                VestaField::from(public_input_val),
+                VestaField::from(output_val)
+            ]
+        );
+
+        // modify the output value
```
Comment (mimoo) [inline src/backends/kimchi/prover.rs:308]: you don't modify anything here

File: src/circuit_writer/mod.rs [hunk 7]
```diff
@@ -223,6 +223,17 @@ impl<B: Backend> CircuitWriter<B> {
...
         let public_output = circuit_writer.public_output.clone();
 
+        // constraint public outputs to the result of the circuit
+        if let Some(public_output) = &public_output {
+            let cvars = &public_output.cvars;
+
+            for (pub_var, ret_var) in cvars.iter().zip(&returned_cells.clone().unwrap()) {
+                circuit_writer
+                    .backend
+                    .assert_eq_var(pub_var.cvar().unwrap(), ret_var, main_span);
```
Comment (mimoo) [inline src/circuit_writer/mod.rs:233]: beautiful 🤩

File: src/backends/kimchi/mod.rs [hunk 8]
```diff
@@ -384,7 +385,24 @@ impl Backend for KimchiVesta {
...
+            for (row, col) in rows_cols {
+                witness[row][col] = val;
+            }
+            public_outputs.push(val);
+        }
+
+        // sanity check the witness
+        for (row, (gate, witness_row, debug_info)) in
+            izip!(self.gates.iter(), &witness, &self.debug_info).enumerate()
+        {
```
Comment (mimoo) [inline src/backends/kimchi/mod.rs:405]: btw what is this change related to in this file?

**Review by mimoo — COMMENTED** [review]
Ah I see! I guess it would be nice to also check that the witness verification function returns an error when we modify the public output then :o 

best practice is to include a screenshot of the test failing _before_ you fix the test by adding the constraints, can you do that :D?

also I didn't review some code that seems unrelated to this PR, so wondering if I need to?

File: src/backends/kimchi/prover.rs (hunk 5 above)
Comment (katat) [reply src/backends/kimchi/prover.rs:323]: it does this check here:
https://github.com/zksecurity/noname/blob/444ca6a2c8f19318f224ed64e93b315c96b938c4/src/backends/kimchi/prover.rs#L325-L341

perhaps we can also make it assert it returns `DisconnectedWires ` error (see the above code snippet) if the output is modified.

File: src/backends/kimchi/mod.rs (hunk 8 above)
Comment (katat) [reply src/backends/kimchi/mod.rs:405]: The `generate_witness` function was assuming there is only a `PublicOutput` var in the circuit. But after constraining the output to the result of circuit, this public output cell will appear twice at least in the circuit:
 - One time as the public output constraint (as part of the public entry), which will be cancelled by the verifier
 - Second time as the part of the constraint with the result of the circuit

So here we need to have a hash map to record where are the places this output cell appears in circuit. With this location info, the output value calculated at the last via `compute_var` can be put to the correct `<row, col>`.

**Comment by katat** [comment]
> best practice is to include a screenshot of the test failing before you fix the test by adding the constraints, can you do that :D?

I rebased the commits to showcase the failed test at the commit https://github.com/zksecurity/noname/pull/59/commits/8ff921e966f743b7464199d3252d6daff0d88ff7

screenshot: 
Screenshot 2024-05-06 at 10 22 34

Then it is fixed by the subsequent commit, which details the test on different types of verification errors.

> also I didn't review some code that seems unrelated to this PR, so wondering if I need to?

I think they are related to the fix. Please refer to my comments above :D

File: src/backends/kimchi/mod.rs [hunk 9]
```diff
@@ -361,19 +360,21 @@ impl Backend for KimchiVesta {
...
-            izip!(self.gates.iter(), &self.rows_of_vars, &self.debug_info).enumerate()
-        {
+        // calculate witness except for public outputs
+        for (row, row_of_vars) in self.rows_of_vars.iter().enumerate() {
             // create the witness row
             let mut witness_row = [Self::Field::zero(); NUM_REGISTERS];
 
             for (col, var) in row_of_vars.iter().enumerate() {
                 let val = if let Some(var) = var {
                     // if it's a public output, defer it's computation
```
Comment (mimoo) [inline src/backends/kimchi/mod.rs:372]: perhaps here we should add a comment "to avoid deep recursion"

File: src/backends/kimchi/mod.rs (hunk 8 above)
Comment (mimoo) [reply src/backends/kimchi/mod.rs:405]: oh right... I was thinking about it like we're using the permutation for `assert_eq`, but we still don't :D (cf https://github.com/zksecurity/noname/issues/55)

File: src/backends/kimchi/mod.rs [hunk 10]
```diff
@@ -361,19 +360,21 @@ impl Backend for KimchiVesta {
 
         let mut witness = vec![];
         // compute each rows' vars, except for the deferred ones (public output)
-        let mut public_outputs_vars: Vec<(usize, CellVar)> = vec![];
+        let mut public_outputs_vars: HashMap<CellVar, Vec<(usize, usize)>> = HashMap::new();
```
Comment (mimoo) [inline src/backends/kimchi/mod.rs:363]: can you use the `Cell` struct instead of `(usize, usize)` here?

File: src/backends/kimchi/mod.rs [hunk 11]
```diff
@@ -361,19 +360,21 @@ impl Backend for KimchiVesta {
 
         let mut witness = vec![];
```
Comment (mimoo) [inline src/backends/kimchi/mod.rs:361]: unrelated: this really needs a `Vec::with_capacity(self.rows_of_vars.len())`

**Review by mimoo — APPROVED** [review]
awesome, makes sense to me now :) thanks for taking the time to explain!
